File: ios/ExpoTotpModule.swift
import ExpoModulesCore
import CommonCrypto

fileprivate let CHANGE_EVENT_NAME = "onTotpUpdate"

public class ExpoTotpModule: Module {
    private var timer: Timer?
    
    public func definition() -> ModuleDefinition {
        Name("ExpoTotp")

        Events(CHANGE_EVENT_NAME)
      
        AsyncFunction("getTotp", getTotp)
        
        AsyncFunction("startUpdates", startUpdates)
        
        AsyncFunction("stopUpdates", stopUpdates)

        AsyncFunction("start", startUpdates)
        
        AsyncFunction("stop", stopUpdates)
        
    }
    
    private func getTotp(secretKey: String, options: TotpOptions?) -> [String: Any]? {
        let secretBase64 = Data(secretKey.utf8).base64EncodedString()
        let finalOptions = options ?? TotpOptions()
        
        return computeTotp(secretBase64: secretBase64, options: finalOptions)
    }
    
    private func startUpdates(secretKey: String, options: TotpOptions?){
        stopUpdates()
        
        let secretBase64 = Data(secretKey.utf8).base64EncodedString()
        
        let finalOptions = options ?? TotpOptions()
        
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            timer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
                guard let self else { return }
                guard let totpInfo = computeTotp(secretBase64: secretBase64, options: finalOptions) else {
                    return
                }
                sendEvent(CHANGE_EVENT_NAME, totpInfo)
            }
        }
    }
    
    private func stopUpdates(){
        timer?.invalidate()
        timer = nil
    }
    
    private func computeTotp(secretBase64: String, options: TotpOptions) -> [String: Any]? {
        let currentTime = Int(Date().timeIntervalSince1970)
        let remainingTime = options.interval - currentTime % options.interval
        let currentInterval = currentTime / options.interval
        
        guard let keyData = Data(base64Encoded: secretBase64) else {
          NSLog("Invalid secret provided")
          return nil
        }
        
        // Create message for HMAC
        var counter = UInt64(currentInterval).bigEndian
        let counterData = withUnsafeBytes(of: &counter) { Data($0) }

        // Generate HMAC-SHA512 hash
        var hash = [UInt8](repeating: 0, count: Int(options.algorithm.getLength()))
        keyData.withUnsafeBytes { keyBytes in
          counterData.withUnsafeBytes { counterBytes in
              CCHmac(CCHmacAlgorithm(options.algorithm.getAlgorithm()), keyBytes.baseAddress, keyData.count, counterBytes.baseAddress, counterData.count, &hash)
          }
        }

        // Extract offset and compute TOTP
        let offset = hash[hash.count - 1] & 0x0F
        let truncatedHash = (UInt32(hash[Int(offset)]) & 0x7F) << 24 |
                            (UInt32(hash[Int(offset) + 1]) & 0xFF) << 16 |
                            (UInt32(hash[Int(offset) + 2]) & 0xFF) << 8 |
                            (UInt32(hash[Int(offset) + 3]) & 0xFF)
        
        let otpMod = UInt32(pow(10.0, Double(options.digits)))
        let otp = truncatedHash % otpMod
        let otpString = String(format: "%0\(options.digits)d", otp)

        return [
          "code": otpString,
          "remainingTime": remainingTime,
          "progress": (Double(remainingTime) / Double(options.interval)) * 100
        ]
    }
}

File: android/src/main/java/expo/modules/totp/ExpoTotpModule.kt
package expo.modules.totp

import android.os.Bundle
import androidx.core.os.bundleOf
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.Timer
import java.util.TimerTask
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import android.util.Base64
import kotlin.math.pow

const val CHANGE_EVENT_NAME = "onTotpUpdate"

class ExpoTotpModule : Module() {

  private var timer: Timer? = null

  override fun definition() = ModuleDefinition {

    Name("ExpoTotp")

    Events(CHANGE_EVENT_NAME)

    AsyncFunction("getTotp", this@ExpoTotpModule::getTotp)

    AsyncFunction("startUpdates", this@ExpoTotpModule::startUpdates)

    AsyncFunction("stopUpdates", this@ExpoTotpModule::stopUpdates)

    AsyncFunction("start", this@ExpoTotpModule::startUpdates)

    AsyncFunction("stop", this@ExpoTotpModule::stopUpdates)

  }

  private fun getTotp(secretKey: String, options: TotpOptions?): Bundle {
    val secretBase64 = Base64.encodeToString(secretKey.toByteArray(), Base64.DEFAULT)
    val finalOptions = options ?: TotpOptions()

    return this@ExpoTotpModule.computeTotp(secretBase64, finalOptions)
  }

  private fun startUpdates(secretKey: String, options: TotpOptions?){
    stopUpdates()

    timer = Timer()

    val secretBase64 = Base64.encodeToString(secretKey.toByteArray(), Base64.DEFAULT)

    val finalOptions = options ?: TotpOptions()

    timer?.schedule(object : TimerTask() {
      override fun run() {
        val totpInfo = this@ExpoTotpModule.computeTotp(secretBase64, finalOptions)
        sendEvent(CHANGE_EVENT_NAME, totpInfo)
      }
    },0,1000)
  }

  private fun stopUpdates(){
    timer?.cancel()
    timer = null
  }

  private fun computeTotp(secretBase64: String, options: TotpOptions): Bundle {
    val currentTime = (System.currentTimeMillis() / 1000).toInt()
    val remainingTime = options.interval - (currentTime % options.interval)
    val currentInterval = currentTime / options.interval

    val keyBytes = Base64.decode(secretBase64, Base64.DEFAULT)
    val message = ByteArray(8) { i -> (currentInterval.toLong() shr (56 - i * 8)).toByte() }

    val mac = Mac.getInstance(options.algorithm.toString())
    mac.init(SecretKeySpec(keyBytes, options.algorithm.toString()))
    val hash = mac.doFinal(message)

    val offset = hash[hash.size - 1].toInt() and 0x0F
    val truncatedHash = hash.copyOfRange(offset, offset + 4)

    val binary = (truncatedHash[0].toInt() and 0x7F shl 24) or
            ((truncatedHash[1].toInt() and 0xFF) shl 16) or
            ((truncatedHash[2].toInt() and 0xFF) shl 8) or
            (truncatedHash[3].toInt() and 0xFF)

    val otpMod = 10.0.pow(options.digits).toInt()

    val otp = (binary % otpMod).toString().padStart(options.digits, '0')

    return bundleOf(
      "code" to otp,
      "remainingTime" to remainingTime,
      "progress" to remainingTime.toDouble() / options.interval * 100,
    )
  }

}
